swarm/src/handler/from_fn.rs

```diff
@@ -0,0 +1,349 @@
+        self.inbound_streams.push(inbound_future);
+
+        if let Some(waker) = self.idle_waker.take() {
+            waker.wake();
+        }
+    }
+
+    fn inject_fully_negotiated_outbound(
+        &mut self,
+        protocol: <Self::OutboundProtocol as OutboundUpgradeSend>::Output,
+        info: Self::OutboundOpenInfo,
+    ) {
+        let outbound_future = (self.on_new_outbound)(protocol, &mut self.state, info);
+        self.outbound_streams.push(outbound_future);
+
+        if let Some(waker) = self.idle_waker.take() {
+            waker.wake();
+        }
+    }
+
+    fn inject_event(&mut self, event: Self::InEvent) {
+        match event {
+            InEvent::UpdateState(new_state) => self.state = new_state,
+            InEvent::NewOutbound(open_info) => {
+                if self.pending_dials.len() >= self.pending_dial_limit {
+                    self.failed_open
+                        .push_back(OpenError::LimitExceeded(open_info));
+                } else {
+                    self.pending_dials.push_back(open_info);
+                }
+            }
+        }
+    }
+
+    fn inject_dial_upgrade_error(
+        &mut self,
+        info: Self::OutboundOpenInfo,
+        error: ConnectionHandlerUpgrErr<<Self::OutboundProtocol as OutboundUpgradeSend>::Error>,
+    ) {
+        match error {
+            ConnectionHandlerUpgrErr::Timeout => {
+                self.failed_open.push_back(OpenError::Timeout(info))
+            }
+            ConnectionHandlerUpgrErr::Timer => {}
+            ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Select(negotiation)) => self
+                .failed_open
+                .push_back(OpenError::NegotiationFailed(info, negotiation)),
+            ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Apply(apply)) => {
+                void::unreachable(apply)
+            }
+        }
+    }
+
+    fn connection_keep_alive(&self) -> KeepAlive {
+        if self.inbound_streams.is_empty()
+            && self.outbound_streams.is_empty()
+            && self.pending_dials.is_empty()
+        {
+            return KeepAlive::No;
+        }
+
+        KeepAlive::Yes
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<
+        ConnectionHandlerEvent<
+            Self::OutboundProtocol,
+            Self::OutboundOpenInfo,
+            Self::OutEvent,
+            Self::Error,
+        >,
+    > {
+        if let Some(error) = self.failed_open.pop_front() {
+            return Poll::Ready(ConnectionHandlerEvent::Custom(OutEvent::FailedToOpen(
+                error,
+            )));
+        }
+
+        if let Some(outbound_open_info) = self.pending_dials.pop_front() {
+            return Poll::Ready(ConnectionHandlerEvent::OutboundSubstreamRequest {
+                protocol: SubstreamProtocol::new(
+                    ReadyUpgrade::new(self.protocol),
+                    outbound_open_info,
+                ),
+            });
+        }
```

Comment:
Following our coding guidelines i think we should move this block to the bottom so that we first return finished local work (finished upgrades on existing substreams) before we open new substreams.